tests/kernel/smp/src/main.c

```diff
@@ -37,45 +52,354 @@ void t2_fn(void *a, void *b, void *c)
 	 * without a separate CPU!), so the main thread can always
 	 * check its progress.
 	 */
+
 	while (1) {
 		k_busy_wait(DELAY_US);
 		t2_count++;
 	}
 }
 
-void test_main(void)
+/**
+ * @brief Verify multi core execution with 2 preemptive threads
+ *
+ * @details Create a thread at a fixed priority lower than the current
+ * thread and greater than test_main thread. So it preempts
+ * test_main. In uniprocessor mode, this thread will never be
+ * scheduled and the test will fail.
+ */
+void test_smp_preemp_threads(void)
 {
-	int i, ok = 1;
+	k_thread_priority_set(k_current_get(), K_PRIO_PREEMPT(1));
 
-	/* Sleep a bit to guarantee that both CPUs enter an idle
-	 * thread from which they can exit correctly to run the main
-	 * test.
+	k_tid_t tid = k_thread_create(&t2, t2_stack, STACK_SIZE, t2_fn,
+				      NULL, NULL, NULL,
+				      K_PRIO_PREEMPT(2), 0, K_NO_WAIT);
+
+	for (int i = 0; i < 10; i++) {
+		/* Wait slightly longer than the other thread so our
+		 * count will always be lower
+		 */
+		k_busy_wait(DELAY_US + (DELAY_US / 8));
+
+		zassert_true(t2_count > i, "T2 thread isn't"
+			     " scheduled to run on other core");
+	}
+	/* Abort the t2_fn thread */
+	k_thread_abort(tid);
+
+	/* At this point of time, 1 core has current thread running,
+	 * and other core has test_main thread running.
+	 * Once current thread is terminated, test_main reports the
+	 * test suite execution pass/fail.
 	 */
-	k_sleep(100);
+}
 
-	/* Create a thread at a fixed priority lower than the main
-	 * thread.  In uniprocessor mode, this thread will never be
-	 * scheduled and the test will fail.
+/**
+ * @brief Verify multi core execution with 2 cooperative threads
+ *
+ * @details Create a cooperative thread, since current thread is
+ * also cooperative, there is no chance that this thread
+ * will get scheduled in uniprocessor mode.
+ *
+ * For ESP32 only (has only 2 cores): Since t2 thread has
```

Comment:
If the test relies on exactly two cores, you should predicate it on CONFIG_MP_NUM_CPUs somehow, otherwise it will break the instant we add x86 SMP.